Bluetooth: Mesh: Fix Config/Health Client async API

michal-narajowski opened the pull request:
The previous PR was not fully baked, so this PR fixes issues by:
- Adding checks when dereferncing pointers stored in params.
- Clearing ack context if a return argument is NULL to not block and
wait for response.
- Updating Health Client Unack API use in mesh shell and tester

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
test/bluetooth/tester: Use Health Client Unack API

Unacknowledged API is now separate from Ack.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Update Health Client API use

Unack API is now separate from Ack.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

michal-narajowski commented:
```
1/3   MESH    MESH/CL/HM/RFS/BV-01-C   PASS            33.654                  
2/3   MESH    MESH/CL/HM/HPS/BV-01-C   PASS            34.744                  
3/3   MESH    MESH/CL/HM/ATS/BV-01-C   PASS            38.072                  
```
```
 1/30   MESH    MESH/CFGCL/KR/BV-01-C          PASS            47.637                  
 2/30   MESH    MESH/CFGCL/KR/BV-02-C          PASS            84.777                  
 3/30   MESH    MESH/CFGCL/KR/BV-03-C          LT2_NOT_AVAILABLE0.0                  
 4/30   MESH    MESH/CFGCL/KR/BV-04-C          LT2_NOT_AVAILABLE0.0                  
 5/30   MESH    MESH/CFGCL/CFG/SNBP/BV-01-C    PASS            36.283                  
 6/30   MESH    MESH/CFGCL/CFG/COMP/BV-01-C    PASS            33.371                  
 7/30   MESH    MESH/CFGCL/CFG/DTTL/BV-01-C    PASS            32.751                  
 8/30   MESH    MESH/CFGCL/CFG/GPXY/BV-01-C    PASS            31.75                  
 9/30   MESH    MESH/CFGCL/CFG/CFGF/BV-01-C    PASS            33.598                  
10/30   MESH    MESH/CFGCL/CFG/CFGR/BV-01-C    PASS            35.326                  
11/30   MESH    MESH/CFGCL/CFG/MP/BV-01-C      PASS            37.246                  
12/30   MESH    MESH/CFGCL/CFG/SL/BV-01-C      PASS            64.324                  
13/30   MESH    MESH/CFGCL/CFG/SL/BV-02-C      PASS            49.466                  
14/30   MESH    MESH/CFGCL/CFG/NKL/BV-01-C     PASS            37.834                  
15/30   MESH    MESH/CFGCL/CFG/NKL/BV-02-C     PASS            33.126                  
16/30   MESH    MESH/CFGCL/CFG/NKL/BV-03-C     PASS            33.53                  
17/30   MESH    MESH/CFGCL/CFG/AKL/BV-01-C     PASS            32.744                  
18/30   MESH    MESH/CFGCL/CFG/AKL/BV-02-C     PASS            36.159                  
19/30   MESH    MESH/CFGCL/CFG/AKL/BV-03-C     PASS            31.972                  
20/30   MESH    MESH/CFGCL/CFG/AKL/BV-04-C     PASS            33.167                  
21/30   MESH    MESH/CFGCL/CFG/MAKL/BV-01-C    PASS            35.324                  
22/30   MESH    MESH/CFGCL/CFG/MAKL/BV-02-C    PASS            37.799                  
23/30   MESH    MESH/CFGCL/CFG/MAKL/BV-03-C    PASS            34.137                  
24/30   MESH    MESH/CFGCL/CFG/MAKL/BV-04-C    PASS            31.928                  
25/30   MESH    MESH/CFGCL/CFG/NID/BV-01-C     PASS            35.289                  
26/30   MESH    MESH/CFGCL/CFG/RST/BV-01-C     PASS            31.755                  
27/30   MESH    MESH/CFGCL/CFG/HBP/BV-01-C     PASS            39.687                  
28/30   MESH    MESH/CFGCL/CFG/HBS/BV-01-C     PASS            37.863                  
29/30   MESH    MESH/CFGCL/CFG/NTX/BV-01-C     PASS            37.805                  
30/30   MESH    MESH/CFGCL/CFG/LPNPT/BV-01-C   PASS            36.867                  
```

PavelVPV reviewed (commented):
Few minor comments:
- Should the API doc be updated? It could say which arguments should be set/unset to make call synchronous/asynchronous.
- It looks a bit overkill to initialize an acknowledged context and then clean it if some return argument is NULL. Perhaps it is possible to avoided?

michal-narajowski commented:
1. Yes, agreed.
2. It can be moved to the top. 

Do you think this async behavior should be moved to a separate API? Or is it ok to keep it as is and add comments in documentation?

PavelVPV commented:
> Do you think this async behavior should be moved to a separate API? Or is it ok to keep it as is and add comments in documentation?

For me it looks ok to use the same functions, but just update documentation. It should not brake backward compatibility in API since previously it was not allowed to pass NULL to the arguments.

Not sure what others think.

alxelax commented:
I really don't get the reason for having command handlers that do nothing if they got NULL instead of expected memory. 
Could you give a little bit more explanation for this?
I agree with @PavelVPV since NULL is the sensible value that changes behavior that should be mentioned in documentation

alxelax commented:
This PR and PR https://github.com/zephyrproject-rtos/zephyr/pull/41028 contradict each other IMO.

michal-narajowski commented:
@alxelax Yes, they contradict each other. 

In the upcoming version of the specification, there is a feature that requires an async behavior of the API. 
For this feature to be complete there would have to be added callbacks for each status message in Config Client. They are already added to Health Client.

michal-narajowski commented:
>     * It looks a bit overkill to initialize an acknowledged context and then clean it if some return argument is NULL. Perhaps it is possible to avoided?

I wanted to do that, but I think that it will hurt readability. cli_prepare() also checks if cli is initialized, so this would have to be called anyway. Then there would be another conditional after that and a conditional at the end. Initializing and clearing ack ctx is not a heavy operation.

trond-snekvik commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -56,6 +56,10 @@ int bt_mesh_cfg_node_reset(uint16_t net_idx, uint16_t addr, bool *status);
  *
+ *  This method can be used asynchronously by setting @p rsp or @p comp
+ *  as NULL. This way the method will not wait for response and will return
+ *  immediately after sending the command.
```
For these two-parameter response functions; I think it would be more sensible to make the function async only if both parameters are NULL. You might want to see that status, but not the resulting phase, for instance

michal-narajowski replied:
Sounds reasonable.

Commit:
Bluetooth: Mesh: Fix Config/Health Client async API

The previous PR was not fully baked, so this PR fixes issues by:
- Adding checks when dereferncing pointers stored in params.
- Clearing ack context if a return argument is NULL to not block and
wait for response.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

alxelax reviewed (changes requested):
LGTM just fix minor typo
